### PR TITLE
Hide the theme without comments.php deprecation message

### LINKED ISSUE #28045: Block-based themes without a comments file trigger a depreciation warning
Block-based Themes throw up a depreciation warning if they do not include a `comments.php` file:

> Deprecated: Theme without comments.php is deprecated since version 3.0.0 with no alternative available. Please include a comments.php template in your theme. in /var/www/html/wp-includes/functions.php on line 5059

This happens whether or not the theme uses Gutenberg's comment blocks. To get around it, the theme can supply a `comments.php` file... but if they do, then the file's contents will be used to render the content inside of of the post contents block. So for example, using an empty placeholder file to suppress the error will wipe out the display of comments. 😞

The rendering of comments.php is tracked as part of https://github.com/WordPress/gutenberg/issues/24101, but I'm opening this separate issue to see if we can just suppress that warning for block-based themes. It seems like the warning will continue to be a problem even after #24101 is fixed.

_This came up originally in https://github.com/WordPress/theme-experiments/pull/157._

### PR DESCRIPTION
**Pull request by youknowriad**
closes #28045 

> Block-based Themes throw up a depreciation warning if they do not include a `comments.php` file:
> 
> > Deprecated: Theme without comments.php is deprecated since version 3.0.0 with no alternative available. Please include a comments.php template in your theme. in /var/www/html/wp-includes/functions.php on line 5059

I'm not very proud of what I did to achieve the removal of this deprecation but at the moment, I don't see any other way to do it. Open to suggestions

**Testing instructions**

 - Have an FSE theme without comments.php file
 - Add a "Post comments" block to any post
 - Check the frontend
 - You shouldn't see the deprecation message.

### CODE REVIEW

File: packages/block-library/src/post-comments/index.php [hunk 1]
```diff
@@ -37,8 +35,15 @@ function render_block_core_post_comments( $attributes, $content, $block ) {
...
 
 	$wrapper_attributes = get_block_wrapper_attributes( array( 'class' => $classes ) );
-	$output             = sprintf( '<div %1$s>', $wrapper_attributes ) . ob_get_clean() . '</div>';
-	return $output;
+	$output             = ob_get_clean();
+
+	// There's a deprecation warning generated by WP Core.
+	// Ideally this deprecation is removed from Core.
+	// In the meantime, this removes it from the output.
+	$regex  = '/Deprecated:((.)*\n){3}/m';
```
Comment (gziolo) [inline packages/block-library/src/post-comments/index.php:43]: Is this deprecation warning always printed in English?

Comment (gziolo) [reply packages/block-library/src/post-comments/index.php:43]: Maybe you could narrow it down to matching `comments.php` and `wp-includes/functions.php`?

**Comment by swissspidy** [comment]
Have you considered adding a temporary filter to core to allow Gutenberg to disable the deprecation message in a nicer way? For example in `locate_template`. ([example precedence](https://core.trac.wordpress.org/changeset/25580))

At some point it will need a core solution anyway once FSE themes take off.

`header.php`, `footer.php`, and `sidebar.php` have similar deprecation warnings in WordPress. What about those?

**Comment by youknowriad** [comment]
@swissspidy yes, my idea was to avoid touching core yet, because FSE is still experimental right and it's unclear whether the block will continue to exist as is or be refactored later to be something closer to the Query block (we might have both)

So if we do keep this in FSE when the landing to Core is close, we should probably just remove the deprecation in Core for fie-themes. (add a check there)

**Comment by youknowriad** [comment]
> header.php, footer.php, and sidebar.php have similar deprecation warnings in WordPress. What about those?

These are not used at in FSE themes, comment.php is used by the Post Comments block.

**Comment by mcsf** [comment]
Looking at `_deprecated_file` in `wp-includes/functions.php`, it seems like it would be possible to:

1. Add an action for [`deprecated_file_included`](https://github.com/WordPress/wordpress-develop/blob/98b6d335644e119480b6da0a762deca590d3a526/src/wp-includes/functions.php#L5027) to capture the `$file` in a stack.
2. Add a filter for [`deprecated_file_trigger_error`](https://github.com/WordPress/wordpress-develop/blob/98b6d335644e119480b6da0a762deca590d3a526/src/wp-includes/functions.php#L5036) which will pop `$file` from the stack and match it against `"comments.php"`. If it matches, the filter returns false.

This isn't very elegant for two reasons:
* The filter doesn't pass any data, so we have to juggle it with the action.
* The compatibility logic in `wp-includes/theme-compat/comments.php` responsible for calling `_deprecated_file` is arguably abusing the hook by [passing a translated string as `$file`](https://github.com/WordPress/wordpress-develop/blob/98b6d335644e119480b6da0a762deca590d3a526/src/wp-includes/theme-compat/comments.php#L11), which means that we can't simply check for `$file == 'comments.php`, but rather do a partial string match and hope that the current locale isn't interfering with the string.

But it should nevertheless work reliably and without having to introduce new hooks or manipulate the server response in `post-comments/index.php`. Am I missing something?

**Comment by gziolo** [comment]
> 2\. Add a filter for [`deprecated_file_trigger_error`](https://github.com/WordPress/wordpress-develop/blob/98b6d335644e119480b6da0a762deca590d3a526/src/wp-includes/functions.php#L5036) which will pop `$file` from the stack and match it against `"comments.php"`. If it matches, the filter returns false.

It might be simpler in terms of readability to disable and enable this filter around `comments_template();` call. It's a temporary solution for a few weeks, right?

**Comment by mcsf** [comment]
> It might be simpler in terms of readability to disable and enable this filter around `comments_template();` call. It's a temporary solution for a few weeks, right?

I'm ok with that :)

**Comment by youknowriad** [comment]
I like the filter's approach. Thanks :)

**Review by gziolo — APPROVED** [review]
🚢

File: packages/block-library/src/post-comments/index.php [hunk 2]
```diff
@@ -26,7 +26,22 @@ function render_block_core_post_comments( $attributes, $content, $block ) {
...
 
 	ob_start();
+	// There's a deprecation warning generated by WP Core.
+	// Ideally this deprecation is removed from Core.
+	// In the meantime, this removes it from the output.
+	add_filter(
+		'deprecated_file_trigger_error',
+		function() {
+			return false;
+		}
```
Comment (mcsf) [inline packages/block-library/src/post-comments/index.php:36]: Minor point, but you can use `__return_false` instead.

File: packages/block-library/src/post-comments/index.php [hunk 3]
```diff
@@ -26,7 +26,22 @@ function render_block_core_post_comments( $attributes, $content, $block ) {
...
+	);
 	comments_template();
+	add_filter(
+		'deprecated_file_trigger_error',
+		function() {
+			return true;
+		}
```
Comment (mcsf) [inline packages/block-library/src/post-comments/index.php:43]: Shouldn't this be a saved reference to whatever filter was there before?

File: packages/block-library/src/post-comments/index.php [hunk 4]
```diff
@@ -25,10 +25,23 @@ function render_block_core_post_comments( $attributes, $content, $block ) {
...
 	ob_start();
+	// There's a deprecation warning generated by WP Core.
+	// Ideally this deprecation is removed from Core.
+	// In the meantime, this removes it from the output.
+	add_filter(
+		'deprecated_file_trigger_error',
+		function() {
+			return false;
+		}
+	);
```
Comment (aristath) [inline packages/block-library/src/post-comments/index.php:37]: Could we not just use `__return_false` here?

```suggestion
	add_filter( 'deprecated_file_trigger_error', '__return_false' );
```

File: packages/block-library/src/post-comments/index.php [hunk 5]
```diff
@@ -25,10 +25,23 @@ function render_block_core_post_comments( $attributes, $content, $block ) {
...
 	comments_template();
+	add_filter(
+		'deprecated_file_trigger_error',
+		function() {
+			return true;
+		}
+	);
```
Comment (aristath) [inline packages/block-library/src/post-comments/index.php:44]: Should probably simplify using `__return_true`
```suggestion
	add_filter( 'deprecated_file_trigger_error', '__return_true' );
```

**Review by aristath — CHANGES_REQUESTED** [review]

File: packages/block-library/src/post-comments/index.php (hunk 2 above)
Comment (aristath) [reply packages/block-library/src/post-comments/index.php:36]: 😆 Looks like we were thinking the same and I missed your comment by a few seconds before I hit the publish button on my review

**Review by aristath — APPROVED** [review]

File: packages/block-library/src/post-comments/index.php (hunk 3 above)
Comment (youknowriad) [reply packages/block-library/src/post-comments/index.php]: Actually found a better way  using remove_filter

Comment (mcsf) [reply packages/block-library/src/post-comments/index.php]: Oh, totally. D'oh. :)

**Comment by mtias** [comment]
Do we have an issue opened to figure out what to do with the comments template more broadly? We should avoid `comments_template` in general. cc @scruffian

**Comment by youknowriad** [comment]
@mtias there's this issue to explore a comments block similar to the query block https://github.com/WordPress/gutenberg/issues/24101

**Comment by mtias** [comment]
Ah, great!

**Pull request merged**